docs/proposals/addon-auth-pattern.md

```diff
@@ -0,0 +1,126 @@
+- Do not use basic-auth creds in browsers, which are subject to CSRF attacks.
+- Do not use client certs in browsers, which are hard to install.
+
+The example used in this pattern is the kubernetes Dashboard.
+
+### Current pattern which has problems
+
+The following pattern is in use today.
+
+When a cluster service, such as Dashboard is being deployed:
+1. Dashboard deployment and service are is created, in namespace `kube-system`.  Service is called `dashboard`.
+
+When a user wants to access a service:
+1. User runs `kubectl proxy -p 8001 &`.
+  -  This causes kubectl to serve HTTP on `localhost:8001`, and to reverse proxy that to the users current cluster, using TLS and the users
+     currently configured authentication method.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.
+  - The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/#/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. The Dashboard sees the request.  It does not know what users is making the request.
+1. The response is sent back from the Pod to the Apiserver.
+1. The response is sent back to the user.
+1. The user sees the Kubernetes Dashboard page.
+
+The problem with this is that the Dashboard does not know which user is accessing it, so it cannot support multiple users with different capabilities.
+
+### Proposed Pattern
+
+The proposed pattern has the following key improvements over the current pattern:
+- a header is injected which tells the Dashboard pod the user who is reaching it.
+- the header can be validated easily as being signed by the apiserver, which is trusted to authenticate users
+
+When a cluster service, such as Dashboard is being deployed:
+1. Dashboard deployment and service are is created, in namespace `kube-system`.  Service is called `dashboard`.
+  - *NEW* The service has annotation `alpha.service-proxy.kubernetes.io/proxy-authentication` set to `true`.
+    Only services that want the authentication header injected set this option.
+
+When a user wants to access a service:
+1. User runs `kubectl proxy -p 8001 &`.
+1. User enters this into their browser navigation bar: `http://localhost:8001/api/v1/proxy/namespaces/kube-system/services/kubernetes-dashboard/#/workload`
+  - This causes the API server to proxy a single HTTP request.
+  - The effect of the two proxies is that an HTTP request on localhost:8001 becomes an HTTP request for `/#/workloads` on port 80 on the pod IP of one of the Pods that backs services `dashboard` in namespace `kube-system`.  So far so good.
+1. *NEW* When the API server proxies a request to a service with annotation `alpha.service-proxy.kubernetes.io/proxy-authentication`, it does the following extra steps:
```

Comment:
it feels odd that the `proxy-authentication` annotation only works when proxy-ing through the API server, would it make sense to implement this lower in the stack (at the same level as kube-proxy for example) such that if the annotation is applied, a `proxy-authn` add-on will do the exact same logic as the proposed solution for the API server.

This way, this annotation is a lot more powerful since it will seamlessly be applied regardless of how the end-user chooses to access a service.